First version of app with basic funcionallity

kubakostrzewski opened the pull request:
Command-line application. Used technologies: Spring boot, Hibernate, MySQL. 
Funcionallity:
- register new user
- login user
- add new task
- show user's tasks list

kubakostrzewski committed:
First commit

* new Spring Boot project
* added hibernate model classes
* added repository classes

kubakostrzewski committed:
Added login, register, addTask, showTask

* added login user funcionality
* added register new user funcionality
* added add new task funcionallity
* added show user tasks list funcionality

kubakostrzewski committed:
Refactor

* added LoginService class
* added TaskService class
* added UserService class

kubakostrzewski committed:
Refactor

* added LoginService class
* added TaskService class
* added UserService class

kubakostrzewski committed:
Merge branch 'kuba-kostrzewski' of https://github.com/kubakostrzewski/StowarzyszenieNaukiJavy into kuba-kostrzewski

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/controller/AppController.java:
```diff
@@ -0,0 +1,95 @@
+@Controller
+public class AppController {
+
+    private static final int UNDEFINED = -1;
```
Wyciągnąłbym te wszystkie stałe do enuma i nazwał np. Operations.

kubakostrzewski replied:
Operacje zostały wyciągnięte do 2 klas enum

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/service/LoginService.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Scanner;
+
+@Service
+public class LoginService {
```
Mieszasz odpowiedzialności - LoginService jest odpowiedzialny za wczytywanie danych od użytkownika (zawsze z konsoli!) i odczyt z bazy. Może warto byłoby ten serwis zostawić tylko dla komunikacji z bazy danych i ewentualnej walidacji (?).

Część odpowiedzialną za wczytywanie danych wyrzuciłbym do osobnej klasy np. LoginController - dzięki temu oddzielimy te dwie logiki.

Mając takie rozgraniczenie możesz później tworzyć beany np. LoginRestController, LoginConsoleController - a LoginService zawsze zostaje taki sam.

kubakostrzewski replied:
Wydzieliłem z klas typu Service klasy typu Controller. Controllery są odpowiedzialne za komunikację z użytkownikiem, a Service za komunikację z bazą. Dodatkowo Controllery są rozszerzeniem interfejsu żeby w przyszłości łatwo podmienić konsolę na coś innego

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/service/TaskService.java:
```diff
@@ -0,0 +1,109 @@
+import java.util.stream.IntStream;
+
+@Service
+public class TaskService {
```
To samo co w przypadku LoginService.

kubakostrzewski replied:
jw

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/service/UserService.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.Scanner;
+
+@Service
+public class UserService {
```
To samo co LoginService oraz TaskService.

kubakostrzewski replied:
jw

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/service/TaskService.java:
```diff
@@ -0,0 +1,109 @@
+        this.userRepository = userRepository;
+    }
+
+    public User addTask(User user){
```
Wiem, że po refaktoryzacji klasy kod będzie całkiem inny, ale przestrzegam, że niezbyt dobrze wygląda metoda addTask, do której przekazujesz obiekt User.

kubakostrzewski replied:
Po refaktoryzacji klasa rzeczywiście wygląda trochę inaczej ale nie wiem czy nie trzeba tego jeszcze poprawić

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/service/TaskService.java:
```diff
@@ -0,0 +1,109 @@
+                    .range(0, userTasks.size())
+                    .forEach(i -> {
+                                Task task = userTasks.get(i);
+                                System.out.println(i + 1 + ".\ttytuł: " + task.getTitle() + "\n" +
```
Może lepiej by było przeciążyć według własnego uznania metodę toString?

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/service/TaskService.java:
```diff
@@ -0,0 +1,109 @@
+    public void showAllTasks(User user){
+        System.out.println("Lista zadań:");
+        List<Task> userTasks = user.getTasks();
+        if (userTasks.size() == 0 || userTasks == null){
```
Co w tym przypadku daje Ci sprawdzenie userTasks == null jako druga część warunku? Na samym początku i tak wywoła się część z size - taka konstrukcja nie uchroni Cię przed NullPointerException.

Ogólnie nie sprawdzałbym tego czy jest nullem - w User inicjalizujesz zawsze listę, a hibernate nie wrzuci Ci tam nulla. No chyba, że gdzieś samemu to zrobisz. ;)

```
userTasks.size() == 0
```
Zamień na:
```
userTasks.isEmpty()
```

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/service/TaskService.java:
```diff
@@ -0,0 +1,109 @@
+        }
+    }
+    private boolean isFieldEmpty(String value, String name){
+        if (value == null || value.equals("")){
```
Do operacji na Stringa warto skorzystać z takiej fajnej biblioteczki: https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/service/TaskService.java:
```diff
@@ -0,0 +1,109 @@
+        if (value == null || value.equals("")){
+            System.out.println("Pole \"" + name + "\" nie może być puste!");
+            return true;
+        } else return false;
```
Nie potrzebny else - nie komplikuj na siłę kodu.

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/service/TaskService.java:
```diff
@@ -0,0 +1,109 @@
+        Date date = null;
+        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm");
+        StringBuilder builder = new StringBuilder();
+        while (date == null){
```
Czy aby na pewno parser zadziała jeśli raz wpiszesz coś źle, a drugie podejście będzie dobre?

Na pierwszy rzut oka wydaje mi się, że powinieneś tworzyć StringBuilder co każde przejście pętli.

kubakostrzewski replied:
Poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/service/UserService.java:
```diff
@@ -0,0 +1,75 @@
+            return true;
+        } else return false;
+    }
+    private boolean checkLogin(String login) {
```
Jeśli przyjmiemy twoje wczytywanie danych jako taki 'frontend' to powinieneś wykonywać tutaj tylko walidację danych typowo frontendowo - np. długość, format etc.

Dopiero, gdy dane pójdą dalej (do serwisu) tam powinny być walidowane czy coś faktycznie istnieje.

kubakostrzewski replied:
poprawione, teraz walidacja tego czy pola są puste jest na poziomie Controllerów, a sprawdzenie poprawności z bazą danych na poziomie Serwisów

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/service/UserService.java:
```diff
@@ -0,0 +1,75 @@
+        } else return false;
+    }
+    private boolean isFieldEmpty(String value, String name){
+        if (value == null || value.equals("")){
```
StringUtils ponownie

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/service/UserService.java:
```diff
@@ -0,0 +1,75 @@
+        System.out.println("Użytkownik został poprawnie zarejestrowany\n" +
+                "Możesz się teraz zalogować");
+    }
+    private boolean checkPassword(String password){
```
Metody z przedrostkiem check nie są dla mnie zrozumiałe na pierwszy rzut oka - muszę się wczytywać kiedy zwracają true, a kiedy false. Myślałem, że zwracają true kiedy jest ok, lecz u Ciebie jest inaczej. Doszedłem do tego dopiero po przeczytaniu kodu, a tak być nie powinno.

Zamień metody np. na isLoginCorrect(String login) lub w drugą strone: incorrect. Będzie to czytelniejsze tak jak isFieldEmpty.

kubakostrzewski replied:
poprawione zostały wszystkie metody sprawdzające poprawność danych

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/test/java/pl/kostrzej/simpleToDoApp/SimpleToDoAppApplicationTests.java:
```diff
@@ -0,0 +1,17 @@
+
+	@Test
+	public void contextLoads() {
+	}
```
Brak jakichkolwiek testów :(
Możliwe, że łatwiej byłoby je napisać, gdyby każda klasa miała jedną odpowiedzialność zamiast kilku.

kubakostrzewski replied:
dodałem 2 testy, kolejne dodam w przyszłych commitach

1024kb-pl reviewed (changes requested):
Ogólnie:
Mieszasz bardzo odpowiedzialności - klasy w jednym momencie wczytują dane od użytkownika, robią walidację nawet z bazą danych, zapisują i wyświetlają efekt operacji na ekran.
Jest tutaj mały misz masz, proponuję wydzielić aplikację na mniejsze (reużywalne) moduły - w tym momencie każda Twoja klasa korzysta z konsoli, a co, gdy będziemy chcieli przenieść jakąś klasę do aplikacji Webowej?

kubakostrzewski committed:
Refaktor kodu

* Rozdzielenie logiki odczytu danych od użytkownika i komunikacji
  z bazą danych między klasy Controller i Service
* Dodanie interfejsów TaskController, UserController, LoginController
  i ich implementacji
* Dodanie klas enum przechowujących opcje menu
* Dodanie klas wyjątków
* Dodanie klasy walidatora danych wpisanych w konsoli i metody sprawdzającej
  czy podana wartość nie jest pusta
* Dodanie klasy walidatora urzytkownika i metodu sprawdzjącej czy
  użytkownik o podanym loginie już istnieje
* Zmieniny podział na pakiety
* Dodanie testów jednostkowych

kubakostrzewski committed:
Refaktor kodu

* Rozdzielenie logiki odczytu danych od użytkownika i komunikacji
  z bazą danych między klasy Controller i Service
* Dodanie interfejsów TaskController, UserController, LoginController
  i ich implementacji
* Dodanie klas enum przechowujących opcje menu
* Dodanie klas wyjątków
* Dodanie klasy walidatora i metody sprawdzającej czy pole jest
  puste
* Zmieniny podział na pakiety
* Dodanie testów jednostkowych

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/user/UserService.java:
```diff
@@ -0,0 +1,24 @@
+    @Autowired
+    private UserRepository userRepository;
+
+    public void addUser(String login, String password, String email, String question, String answer){
```
Mając metodę w serwisie lepiej będzie przekazać już gotowy obiekt - w tym przypadku User. Przypuśćmy, że mamy Controller, który otrzymuje Usera w postaci JSON - chcąc wywołać metodę addUser będziesz musiał rozbić już gotowego Usera, żeby przekazać jego pola jako argumenty.

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/user/UserService.java:
```diff
@@ -0,0 +1,24 @@
+    private UserRepository userRepository;
+
+    public void addUser(String login, String password, String email, String question, String answer){
+        if (!userRepository.existsByLogin(login)){
```
Dobrym pomysłem byłoby stworzyć klasę UserValidatorService walidującą usera. Możesz w niej zawrzeć konkretne metody lub listę walidatorów (klas implementujących np. interfejs UserValidator), gdzie każda z osobnych klas byłaby odpowiedzialna za sprawdzanie czego innego.

1024kb-pl replied:
Wtedy klasa UserValidatorService mogłaby mieć metodę void validate, która rzucałaby wyjątkiem z odpowiednią wiadomość w przypadku, gdy coś jest nie tak z userem.

kubakostrzewski replied:
dodałem klasę UserValidator

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/validator/FieldValidator.java:
```diff
@@ -0,0 +1,15 @@
+
+    public boolean isFieldEmpty(String value, String name){
+        if (value.isEmpty()){
+            System.out.println("Pole \"" + name + "\" nie może być puste!");
```
Komponent FieldValidator jest używalny tylko w środowisku konsoli - no chyba, że chcesz tylko logować informację o nie przejściu walidacji, ale w takim przypadku powinieneś wykorzystać Logger.

kubakostrzewski replied:
Chodzi o validację po stronie "front endu" czy wartość spełnia wymagania, w tym przypadku czy wymagane pole nie jest puste, później chciałem dodać też np sprawdzenie czy adres email jest w odpowiednim formacie itp

kubakostrzewski committed:
Merge branch 'kuba-kostrzewski' of https://github.com/kubakostrzewski/StowarzyszenieNaukiJavy into kuba-kostrzewski

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/validator/UserValidator.java:
```diff
@@ -0,0 +1,18 @@
+    @Autowired
+    private UserRepository userRepository;
+
+    public void validate(User user){
```
Na przyszłość: w przypadku, gdybyś nie pisał frontu (tak jak u Ciebie to wygląda w kontrollerze) to pamiętaj w Validatorze sprawdzać również czy pola nie są puste, mają minimalną lub maksymalną długość. Są to takie podstawowe case.

kubakostrzewski replied:
Mimo, że walidacja jest w kontrollerze to dodałem dodatkową walidację w metodzie validate

kubakostrzewski replied:
Dodałem mimo wszystko walidację wszystkich pól usera

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/app/MainMenuOptions.java:
```diff
@@ -0,0 +1,26 @@
+        this.name = name;
+        this.number = number;
+    }
+    static MainMenuOptions returnIfCorrect(int number){
```
Metodę uzależniasz od podanych numerów - sprawdzając zakresy. Lepiej by było przeiterować po values() i sprawdzić czy jest tam podana wartość np. przy użyciu filter. Wszystko na koniec można zapakować w optionala  i użyć orElseThrow.

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/app/WelcomeMenuOptions.java:
```diff
@@ -0,0 +1,26 @@
+        this.number = number;
+        this.name = name;
+    }
+    static WelcomeMenuOptions returnIfCorrect(int number){
```
Tak samo jak w drugim enumie od opcji.

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/login/LoginService.java:
```diff
@@ -0,0 +1,19 @@
+    private UserRepository userRepository;
+
+    public User logIn(String login, String password){
+        User user = userRepository.findByLogin(login);
```
Wykorzystaj Optionala w połączeniu z orElseThrow - wystarczy, że w userRepository zmienisz typ zwracany User na Optional<User>.

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/login/LoginService.java:
```diff
@@ -0,0 +1,19 @@
+
+    public User logIn(String login, String password){
+        User user = userRepository.findByLogin(login);
+        if (user != null && user.getPassword().equals(password)) return user;
```
Jeśli użyjesz Optionala nie będziesz musiał sprawdzać nulla.

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/login/LoginService.java:
```diff
@@ -0,0 +1,19 @@
+    public User logIn(String login, String password){
+        User user = userRepository.findByLogin(login);
+        if (user != null && user.getPassword().equals(password)) return user;
+        else throw new InvalidLoginDataException();
```
Else jest już zbędny.

kubakostrzewski replied:
dodałem logowanie przed rzuceniem wyjątku w związku z tym else został

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/task/TaskConsoleController.java:
```diff
@@ -0,0 +1,102 @@
+        System.out.println("Podaj opis lub zostaw puste:");
+        description = scanner.nextLine();
+        date = readDate();
+        done = readDone();
```
Z biznesowego punktu widzenia wydaje mi się, że użytkownik nie powinien mieć wyboru podczas tworzenia czy zadanie jest wykonane czy nie - jakby było wykonane to może nie byłoby sensu go tam umieszczać. Domyślnie niech każde zadanie będzie nie wykonane.

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/task/Task.java:
```diff
@@ -0,0 +1,97 @@
+    private String description;
+    @Temporal(TemporalType.TIMESTAMP)
+    private Date date;
+    private boolean done;
```
Czytelniejsze byłoby użycie enuma np. TaskStatus - na początku co prawda mamy 2 stany co opęka nam boolean, ale co w przypadku, gdy klient zażyczy sobie nowego statusu np. PENDING?

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/task/TaskController.java:
```diff
@@ -0,0 +1,10 @@
+
+import pl.kostrzej.simpleToDoApp.components.user.User;
+
+public interface TaskController {
```
Co z edycją tasków i oznaczaniem ich jako wykonanych?

kubakostrzewski replied:
To zostanie dodane w kolejnych commitach, na razie chciałem wrzucić pierwszą wersję z kilkoma podstawowymi funkcjonalnościami żeby dostać feedback czy idę w dobrym kierunku i co można poprawić

kubakostrzewski replied:
Dodałem ww funkcjonalności

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/user/UserConsoleController.java:
```diff
@@ -0,0 +1,69 @@
+        if (!password.equals(scanner.nextLine())){
+            System.out.println("Podane hasła są różne!");
+            return false;
+        } else return true;
```
Else jest zbędny

kubakostrzewski replied:
dodałem logowanie przed zwróceniem true w związku z tym else został

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/test/java/pl/kostrzej/simpleToDoApp/app/MainMenuOptionsTest.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Test(expected = InvalidOptionException.class)
+    public void choseTooHighOption(){
+        MainMenuOptions.returnIfCorrect(4);
```
Uzależniłbym ten test od MainMenuOptions.values().length + 1. Dodając nowe opcję nie będziesz musiał poprawiać testów.

kubakostrzewski replied:
Wydaje mi się, że zmiana wartości w teście powinna być świadoma i niezależna od klasy, która jest testowana. W przypadku gdy np. niechcący usunę jedną opcję to testy przejdą pozytywnie, a tak na prawdę nie powinny bo opcja, która jest używana w programie została usunięta.

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/test/java/pl/kostrzej/simpleToDoApp/app/WelcomMenuOptionsTest.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Test(expected = InvalidOptionException.class)
+    public void choseTooHighOption(){
+        MainMenuOptions.returnIfCorrect(4);
```
jw

kubakostrzewski replied:
jw ;)

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/task/TaskService.java:
```diff
@@ -0,0 +1,33 @@
+        task.setDate(date);
+        task.setDate(date);
+        taskRepository.save(task);
+        return userRepository.findByLogin(user.getLogin());
```
metoda save zwraca Ci już gotową encję - nie musisz wyszukiwać.

kubakostrzewski replied:
Rzeczywiście, tylko metoda addTask zwraca encję User, a save jest robione dla encji Task. Chodziło mi o to żeby po zapisaniu Taska został wykonany SELECT do bazy i zwrócony ten sam User ale ze zaktualizowaną listą Tasków.

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/task/Task.java:
```diff
@@ -0,0 +1,97 @@
+
+@Entity
+@Table(name = "tasks")
+public class Task implements Serializable {
```
Zainteresuj się Lombokiem + plugin do IntelliJ.

kubakostrzewski replied:
dodałem lomboka

1024kb-pl reviewed (commented):
Projekt już ma trochę klas - może warto by było użyć loggera np. slf4j i logować różnymi poziomami logów w serwisach?

kubakostrzewski committed:
Dodanie loggera, lomboka i refaktor kodu

* dodanie Slf4j i log4j
* dodanie lomboka i zaimplementowanie w klasach encji
* refaktor kodu

kubakostrzewski committed:
Dodana funkcjonalność usuwania zadania

* dodanie funkcjonalności usuwania zadania przez użytkownika
* dodanie klasy enum ze statusami zadania
* dodanie metody pobierającej użytkownika z bazy

kubakostrzewski committed:
Dodanie funkcjonalności ziany statusu zadania

kubakostrzewski committed:
Zmiana interfejsów TaskController i UserController

* Zmiana typów argumentów i wartości zwracanyc
* TaskController przyjmuje argumenty typu Task
* UserController przyjmuje argumenty typu User i zwraca User

kubakostrzewski committed:
Dodanie edycji zadania, refaktor, naprawa bugów

* dodanie funkcjonalności edycji zadanie
* zastąpienie konstruktorów adnotacjami @AllArgsConstructor
* dodanie metody odpowiedzialnej za wybranie zadania z listy
* naprawa bugów związanych z wprowadzaniem nieporawidłowych wartości w menu
* wydzielenie odczytu daty do odzielnej klasy
* zmiana typów argumentów przekazywanych do metod operujących na encjach Task

1024kb-pl commented:
W piątek będę miał dostęp do laptopa to juz zerkne na cały kod i zobaczę jak to się komponuje. ;)

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/model/User.java:
```diff
@@ -0,0 +1,107 @@
+    private String login;
+    @Column(nullable = false)
+    private String password;
+    private String email;
```
czy email nie powinien być też unikalny?;)

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/app/AppController.java:
```diff
@@ -0,0 +1,156 @@
+            printWelcomeMenu();
+            switch (chooseWelcomeMenuOption()) {
+                case LOGIN:
+                    user = loginController.logIn();
```
logIn zwraca usera albo nulla - nie masz rozróżnienia poprawnego logowania od błędnego. W przypadku, gdy logowanie się odbędzie niepoprawnie nie możesz zwracać nulla.

kubakostrzewski replied:
Hmm, jak mógłbym to w takim razie rozwiązać? Teraz gdy logowanie się nie odbędzie i użytkownik będzie nullem to program zostaje w pętli związanej z logowaniem/rejestracją i nie przejdzie do pętli z menu głównym dopóki logowanie nie zwróci poprawnego użytkownika.

1024kb-pl replied:
Przyjrzałem się bliżej i faktycznie operujesz tutaj na null. Nie do końca mi się to podoba, ale niech będzie. ;)

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/app/AppController.java:
```diff
@@ -0,0 +1,156 @@
+                    log.info("Logging user finished. User: " + user);
+                    break;
+                case REGISTER:
+                    userController.addUser();
```
Podobnie jak podczas logowania - wyświetlasz informację, że użytkownik już istnieje, a później, że rejestracja przebiegła pomyślnie. Pomysł o wykorzystaniu wyjątków.

kubakostrzewski replied:
chodzi o to, że po `userValidator.validate(user);` jest od razu `userRepository.save(user);` ? metoda `validate(user)` rzuca wyjątki, które są wyłapywane po stronie `UserController` i jeżeli wyjątek zostanie rzucony to metoda `save(user)` i logowanie nie zostaną wywołane. Czy może o coś innego chodziło?

1024kb-pl replied:
Tak dokładnie o to chodziło - w walidacji jeśli coś pójdzie nie tak to powinien zostać rzucony wyjątek, który powinien przerwać dodawanie użytkownika.

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/task/TaskService.java:
```diff
@@ -0,0 +1,43 @@
+    UserRepository userRepository;
+
+
+    public void addTask(User user, String title, String description, Date date){
```
Lepiej by było, gdyby taskService otrzymywał już stworzony obiekt Task.

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/validator/UserValidator.java:
```diff
@@ -0,0 +1,39 @@
+        log.info("User validation process initialized.");
+        if (user.getLogin().isEmpty()) {
+            log.info("Login is empty.");
+            throw new NullPointerException("Login nie może być pusty!");
```
Nie jest dobrmy pomysłem rzucanie NPE - jest to rzadko widywany wyjątek i nie ma co straszyć innych programistów ich wystąpieniem. Lepiej wykorzystać stworzone przez siebie wyjątki,.

kubakostrzewski replied:
poprawione na `IllegalArgumentException`

1024kb-pl reviewed (changes requested):
Co do logowania masz różne poziomy logowania:

- info - warto umieszczać w punktach wejściowych metod w celach łatwiejszego znajdowania błędu, warto wyświetlać stan obiektów wchodzących jako parametry
- error - oczywiście w przypadku błędów, np. w blokach catch
- warn - coś na co warto zwrócić uwagę
- debug - informację pokazywane tylko w trybe debug, mogą to być jakieś dokładniejsze informację przydatne podczas tworzenia aplikacji

Zamiast dodawania stringów możesz korzystać z {} - a ich wypełnienie podawać po przecinku, np.:
```
log.info("Creating task: {}", task);
```

kubakostrzewski commented:
> Co do logowania masz różne poziomy logowania:
> 
> * info - warto umieszczać w punktach wejściowych metod w celach łatwiejszego znajdowania błędu, warto wyświetlać stan obiektów wchodzących jako parametry
> * error - oczywiście w przypadku błędów, np. w blokach catch
> * warn - coś na co warto zwrócić uwagę
> * debug - informację pokazywane tylko w trybe debug, mogą to być jakieś dokładniejsze informację przydatne podczas tworzenia aplikacji
> 
> Zamiast dodawania stringów możesz korzystać z {} - a ich wypełnienie podawać po przecinku, np.:
> 
> ```
> log.info("Creating task: {}", task);
> ```

Czyli np jak użytkownik poda błędne dane logowania i rzucam `InvalidLoginDataException`, to lepiej logować na poziomie INFO czy ERROR?

kubakostrzewski committed:
Refaktor kodu

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/.gitignore:
```diff
@@ -0,0 +1,27 @@
+.sts4-cache
+
+### IntelliJ IDEA ###
+.idea
```
Masz zadeklarowane .idea, ale nadal jest ono w repo. Musisz usunąć go usunąć komendą git rm.

PS Folder .mvn też nie jest do niczego potrzebny.

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/app/DateController.java:
```diff
@@ -0,0 +1,47 @@
+@Slf4j
+public class DateController {
+
+    @Autowired
```
Stosuj wstrzykiwanie poprzez konstruktor - usuń adnotację i dodaj lomboka i będzie git. ;)

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/app/DateController.java:
```diff
@@ -0,0 +1,47 @@
+                log.info("Date format is valid.");
+            } catch (ParseException e){
+                System.out.println("Podane dane są niewłaściwe!");
+                log.info("Date format is invalid. {}", e.getClass());
```
Poziom logowania info dotyczy informowania tylko, tutaj otrzymujesz błąd - lepiej zastosować error.

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/login/LoginConsoleController.java:
```diff
@@ -0,0 +1,43 @@
+            log.info("Login successful.");
+            return user;
+        } catch (InvalidLoginDataException e){
+            log.info("Invalid login data. {}", e.getClass());
```
Poziom logowania

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/login/LoginService.java:
```diff
@@ -0,0 +1,32 @@
+public class LoginService {
+
+    @Autowired
+    private UserRepository userRepository;
```
Autowiring by constructor

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/login/LoginService.java:
```diff
@@ -0,0 +1,32 @@
+            log.info("Authorization finished successfully.");
+            return user;
+        }
+        else {
```
Zbędny else

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/task/TaskConsoleController.java:
```diff
@@ -0,0 +1,148 @@
+    public Task getTaskFromList(List<Task> tasks){
+        Integer taskIndex = null;
+        showAllTasks(tasks);
+        if (tasks.size()>0){
```
Użyj metody isEmpty()

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/task/TaskConsoleController.java:
```diff
@@ -0,0 +1,148 @@
+                System.out.println("Wpisano niewłaściwe dane! Podaj liczbę!");
+            } catch (IndexOutOfBoundsException e){
+                System.err.println(e.getMessage());
+                log.info("Invalid task number! Task list size= {} Value typed to console = {} {}",
```
Poziom logowania

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/task/TaskConsoleController.java:
```diff
@@ -0,0 +1,148 @@
+    public void showAllTasks(List<Task> tasks){
+        log.info("Show all tasks process initialized.");
+        System.out.println("Lista zadań:");
+        if (tasks.isEmpty() || tasks == null){
```
Nawet jeśli tasks będzie nullem to i tak otrzymasz NPE ze względu, że najpierw wykorzystujesz isEmpty.

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/task/TaskConsoleController.java:
```diff
@@ -0,0 +1,148 @@
+        int taskIndex = scanner.nextInt() - 1;
+        scanner.nextLine();
+        log.info("Correct data type.");
+        if (taskIndex<tasks.size() || taskIndex < 0) {
```
Masz zły warunek - na głos sobie przeczytaj. ;)

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/task/TaskService.java:
```diff
@@ -0,0 +1,30 @@
+@AllArgsConstructor
+public class TaskService {
+
+    TaskRepository taskRepository;
```
Dodaj modyfikator private

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/validator/UserValidator.java:
```diff
@@ -0,0 +1,44 @@
+public class UserValidator {
+
+    @Autowired
+    private UserRepository userRepository;
```
Autowiring by constructor

kubakostrzewski replied:
poprawione

1024kb-pl commented on kuba-kostrzewski/simpleToDoApp/src/main/java/pl/kostrzej/simpleToDoApp/components/validator/UserValidator.java:
```diff
@@ -0,0 +1,44 @@
+    private UserRepository userRepository;
+
+    public void validate(User user){
+        log.info("User validation process initialized.");
```
Zbędne są logi w ifach, wystarczy Ci na początku, gdzie wyświetlisz całego usera. ;)

kubakostrzewski replied:
poprawione

1024kb-pl reviewed (changes requested):
Mało znaczące błędy. ;)

Commit:
Refaktor
